### ISSUE TITLE
Dependency Resolution picks an older version (depending on the order of dependency declarations)

### ISSUE THREAD

**Issue by Adrodoc** [gradle/gradle#7317]
### Expected Behavior
The order of dependency declarations should not affect which versions are used.
In case of a version conflict the newer version should be picked by default.

### Current Behavior
In some cases the order of dependency declarations can influence the dependency resolution in such a way that a version conflict of a transitive dependency is resolved by picking the older of the two versions.

### Steps to Reproduce (for bugs)
With the following `build.gradle` file:
```gradle
apply plugin: 'java'

repositories {
  mavenCentral()
}

dependencies {
  compile 'junit:junit:4.11'
  compile 'pl.pragmatists:JUnitParams:1.0.5'
  compile 'org.mockito:mockito-core:1.9.5'
}
```
Running `gradlew dependencies --configuration compile` gives this output:
```
C:\directory>gradlew dependencies --configuration compile

> Task :dependencies

------------------------------------------------------------
Root project
------------------------------------------------------------

compile - Dependencies for source set 'main' (deprecated, use 'implementation' instead).
+--- junit:junit:4.11 -> 4.12
|    \--- org.hamcrest:hamcrest-core:1.3 -> 1.1
+--- pl.pragmatists:JUnitParams:1.0.5
|    \--- junit:junit:4.12 (*)
\--- org.mockito:mockito-core:1.9.5
     +--- org.hamcrest:hamcrest-core:1.1
     \--- org.objenesis:objenesis:1.0

(*) - dependencies omitted (listed previously)

A web-based, searchable dependency report is available by adding the --scan option.

BUILD SUCCESSFUL in 0s
1 actionable task: 1 executed
```

As you can see this will make use of hamcrest `1.1`, I expected to get hamcrest `1.3`.
Any other order of dependencies seems to cause Gradle to pick the correct version (hamcrest `1.3`).

For example this `build.gradle` file:
```gradle
apply plugin: 'java'

repositories {
  mavenCentral()
}

dependencies {
  compile 'junit:junit:4.11'
  compile 'org.mockito:mockito-core:1.9.5'
  compile 'pl.pragmatists:JUnitParams:1.0.5'
}

```
Results in this output:
```
C:\directory>gradlew dependencies --configuration compile

> Task :dependencies

------------------------------------------------------------
Root project
------------------------------------------------------------

compile - Dependencies for source set 'main' (deprecated, use 'implementation' instead).
+--- junit:junit:4.11 -> 4.12
|    \--- org.hamcrest:hamcrest-core:1.3
+--- org.mockito:mockito-core:1.9.5
|    +--- org.hamcrest:hamcrest-core:1.1 -> 1.3
|    \--- org.objenesis:objenesis:1.0
\--- pl.pragmatists:JUnitParams:1.0.5
     \--- junit:junit:4.12 (*)

(*) - dependencies omitted (listed previously)

A web-based, searchable dependency report is available by adding the --scan option.

BUILD SUCCESSFUL in 0s
1 actionable task: 1 executed
```

### Your Environment

```
C:\directory>gradlew -v

------------------------------------------------------------
Gradle 4.10.2
------------------------------------------------------------

Build time:   2018-09-19 18:10:15 UTC
Revision:     b4d8d5d170bb4ba516e88d7fe5647e2323d791dd

Kotlin DSL:   1.0-rc-6
Kotlin:       1.2.61
Groovy:       2.4.15
Ant:          Apache Ant(TM) version 1.9.11 compiled on March 23 2018
JVM:          1.8.0_162 (Oracle Corporation 25.162-b12)
OS:           Windows 7 6.1 amd64
```

 * Build scan URL (of the first example): https://scans.gradle.com/s/w2tqktzcmj3h4

**Comment by dansanduleac** [comment]
This might have the same underlying cause as https://github.com/gradle/gradle/issues/7050

**Comment by ljacomet** [comment]
I confirm that the fix for #7050 fixes this as well, so closing as duplicate.
